### PR TITLE
feat: changes required to add an extra type detection head to pvrcnn++

### PR DESCRIPTION
**Pull request by julianschoep**
Part of our effort to have pvrcnn++ predict element types.

We're adding an extra classification head to the anchorhead.
This requires some changes to the target proposal logic and post processing logic.

### CODE REVIEW

File: pcdet/models/dense_heads/anchor_head_single.py [hunk 1]
```diff
@@ -1,12 +1,23 @@
+from typing import Iterable, List
+
 import numpy as np
 import torch.nn as nn
 
 from .anchor_head_template import AnchorHeadTemplate
 
 
 class AnchorHeadSingle(AnchorHeadTemplate):
```
Comment (lucagrementieri) [inline pcdet/models/dense_heads/anchor_head_single.py:9]: The class `AnchorHeadSingle` is currently used in our model, so all the proposed changes could break the current code. In particular the addition of a layer can give errors in the loading of the model, but many other problems can arise. This would impact very negatively the reproducibility of the code in my opinion and the rollback in case of problems. It also hinders the possibility of training a model without a second head, something we could want to do in the future.

For all the following reasons, I'm strongly against this breaking change.

I propose to create a completely new class (that could inherit from `AnchorHeadSingle` if needed) to add the support for multiple heads. All the following changes should support both `AnchorHeadSingle` and the new class.

I'm also wondering if [`AnchorHeadMulti`](https://github.com/promaton/OpenPCDet/blob/master/pcdet/models/dense_heads/anchor_head_multi.py) could be used for our purpose given its name 🤔

**Review by lucagrementieri — CHANGES_REQUESTED** [review]

File: pcdet/models/dense_heads/anchor_head_single.py (hunk 1 above)
Comment (julianschoep) [reply pcdet/models/dense_heads/anchor_head_single.py:9]: Hmm good point, although ofcourse this change will be for all products (ortho also needs type detection I guess).
I think it may be very hard to make this feature backwards compatible tbh, I'm going to try to see what we can do but there's a bit of a balance between the value of having it backward compatible and the effort to make it so.
I'm worried it's (much) more work than just adding a new anchor head class, but let me see.

FYI AnchorHeadMulti is a class to have a separate head per class, it does not allow multiple classes per element as we do here.

Comment (julianschoep) [reply pcdet/models/dense_heads/anchor_head_single.py]: I've made it backwards compatible, you can now run the old detection model just by changing the `DENSE_HEAD.NAME` parameter in `pv-rcnn-plusplus_cfg.yaml`

File: pcdet/models/dense_heads/point_head_simple.py [hunk 2]
```diff
@@ -64,7 +64,7 @@ def forward(self, batch_dict):
                 point_features_before_fusion: (N1 + N2 + N3 + ..., C)
                 point_coords: (N1 + N2 + N3 + ..., 4) [bs_idx, x, y, z]
                 point_labels (optional): (N1 + N2 + N3 + ...)
-                gt_boxes (optional): (B, M, 8)
+                gt_boxes (optional): (B, M, 7+N_Class_Heads)
```
Comment (bugra-yilmaz) [inline pcdet/models/dense_heads/point_head_simple.py:67]: `[nit]`
```suggestion
                gt_boxes (optional): (B, M, 7 + N_Class_Heads)
```

File: pcdet/models/dense_heads/point_head_template.py
Comment (bugra-yilmaz) [inline pcdet/models/dense_heads/point_head_template.py:1]: `[curious]` where does number `7` come from?

**Review by bugra-yilmaz — COMMENTED** [review]
Do I understand this correctly, we are actually aiming for using a single detection model to draw bounding boxes around scan elements + predicting element types, so the ement type detection doesn't have its own model?

File: pcdet/models/dense_heads/point_head_template.py
Comment (julianschoep) [reply pcdet/models/dense_heads/point_head_template.py:1]: >Do I understand this correctly, we are actually aiming for using a single detection model to draw bounding boxes around scan elements + predicting element types, so the ement type detection doesn't have its own model?

Yes indeed

File: pcdet/models/dense_heads/anchor_head_template.py [hunk 3]
```diff
@@ -101,38 +101,53 @@ def assign_targets(self, gt_boxes):
...
         one_hot_targets = one_hot_targets[..., 1:]
-        cls_loss_src = self.cls_loss_func(cls_preds, one_hot_targets, weights=cls_weights)  # [N, M]
-        cls_loss = cls_loss_src.sum() / batch_size
 
-        cls_loss = cls_loss * self.model_cfg.LOSS_CONFIG.LOSS_WEIGHTS['cls_weight']
-        tb_dict = {
-            'rpn_loss_cls': cls_loss.item()
-        }
-        return cls_loss, tb_dict
+        return one_hot_targets
```
Comment (lucagrementieri) [inline pcdet/models/dense_heads/anchor_head_template.py:150]: In this file you did a quite large refactor. It's quite difficult to see what is different with respect to the previous version in the diff and to spot errors. In general I'm against large modifications to a fork that we barely maintain because I would like to minimise the number of differences with respect to the main trunk, in particular in files containing important logic.

Can you rewrite the code minimising the number of lines changed?

File: pcdet/models/dense_heads/point_head_template.py
Comment (bugra-yilmaz) [reply pcdet/models/dense_heads/point_head_template.py:1]: > [cx,cy,cz,h,w,d,direction], used to describe a bounding box
It may be better to define this number as a constant somewhere too, to refer to it in docstrings and while indexing the labels.

File: pcdet/models/dense_heads/point_head_template.py [hunk 4]
```diff
@@ -99,7 +99,7 @@ def assign_stack_targets(self, points, gt_boxes, extend_gt_boxes=None,
                 raise NotImplementedError
 
             gt_box_of_fg_points = gt_boxes[k][box_idxs_of_pts[fg_flag]]
-            point_cls_labels_single[fg_flag] = 1 if self.num_class == 1 else gt_box_of_fg_points[:, -1].long()
+            point_cls_labels_single[fg_flag] = 1 if self.num_class == 1 else gt_box_of_fg_points[:, 7].long()
```
Comment (lucagrementieri) [inline pcdet/models/dense_heads/point_head_template.py:102]: Isn't `self.num_class` always a list now? If that's the case this condition will always be false. 

I guess many conditions can be automatically False because of that. Have you checked all the occurrences of `.num_class`=

**Review by bugra-yilmaz — APPROVED** [review]
All good from my side, didn't dive into the model code. Happy to see that we can use the same model for both tasks.

File: pcdet/models/dense_heads/anchor_head_single.py [hunk 5]
```diff
@@ -72,4 +73,10 @@ def forward(self, data_dict):
             data_dict['batch_box_preds'] = batch_box_preds
             data_dict['cls_preds_normalized'] = False
 
-        return data_dict
+            # not predicting type, predict first class for everything
+            B, num_anchors, C = batch_cls_preds.shape
+            type_preds = torch.zeros((B, num_anchors, self.num_class[1]), dtype=batch_cls_preds.dtype, device=batch_cls_preds.device)
```
Comment (lucagrementieri) [inline pcdet/models/dense_heads/anchor_head_single.py:78]: I'm not very convinced by the change of `self.num_class` to a list: I think it's dangerous and in the PR there's already a mistake caused by this change.

You could preserve `self.num_class` as an integer and introduce an additional parameter `num_aux_class` in `AnchorHeadWithType`.

Here you could just use `0` instead of `self.num_class[1]` to create the tensor `data_dict['batch_cls_type_preds']`.

**Review by lucagrementieri — COMMENTED** [review]
I still see dangers in this update.

File: pcdet/models/dense_heads/anchor_head_template.py (hunk 3 above)
Comment (julianschoep) [reply pcdet/models/dense_heads/anchor_head_template.py:150]: I've already done that, and this is the best I can do. It moves some logic from the loss function into separate functions so they can be re-used in `AnchorHeadWithType`.

Comment (julianschoep) [reply pcdet/models/dense_heads/anchor_head_template.py]: Actually you're right, I took over code from Nora but she divided more functions than really necessary. Fixed it now.

File: pcdet/models/dense_heads/point_head_template.py (hunk 4 above)
Comment (julianschoep) [reply pcdet/models/dense_heads/point_head_template.py:102]: Changed it

File: pcdet/models/dense_heads/anchor_head_single.py (hunk 5 above)
Comment (julianschoep) [reply pcdet/models/dense_heads/anchor_head_single.py]: Yeah this was also bugging me, it was a quick solution but indeed a dangerous one, thanks for raising it, I've now added `self.num_aux_class` which is only passed to the anchor head.

File: pcdet/models/dense_heads/target_assigner/axis_aligned_target_assigner.py [hunk 6]
```diff
@@ -38,24 +38,33 @@ def assign_targets(self, all_anchors, gt_boxes_with_classes):
...
         bbox_targets = []
         cls_labels = []
         reg_weights = []
+        aux_cls_labels = []
 
         batch_size = gt_boxes_with_classes.shape[0]
-        gt_classes = gt_boxes_with_classes[:, :, -1]
-        gt_boxes = gt_boxes_with_classes[:, :, :-1]
+        gt_classes = gt_boxes_with_classes[:, :, 7]
+        gt_aux_classes = gt_boxes_with_classes[:, :, 8]
```
Comment (julianschoep) [inline pcdet/models/dense_heads/target_assigner/axis_aligned_target_assigner.py:58]: Note that this does require users to provide a auxilliary label in the gt bbox to the model, even if aux class is not predicted/ used.
Alternatively I'd have to add if-else statements everywhere which seemed like overkill. You can still train without using the aux class though if you want and can populate the aux label with zeros or something

File: setup.py [hunk 7]
```diff
@@ -28,7 +28,7 @@ def write_version_to_file(version, target_file):
 
 
 if __name__ == '__main__':
-    version = '0.6.1+%s' % get_git_commit_number()
+    version = '0.6.2+%s' % get_git_commit_number()
```
Comment (julianschoep) [inline setup.py:31]: Do you think this should be a minor or major version bump?

File: pcdet/models/dense_heads/anchor_head_single.py
Comment (lucagrementieri) [inline pcdet/models/dense_heads/anchor_head_single.py:1]: I guess you can revert all the changes in this file.

File: pcdet/models/dense_heads/anchor_head_template.py [hunk 8]
```diff
@@ -101,14 +101,24 @@ def assign_targets(self, gt_boxes):
     def get_cls_layer_loss(self):
         cls_preds = self.forward_ret_dict['cls_preds']
         box_cls_labels = self.forward_ret_dict['box_cls_labels']
+        cls_loss = self.get_cls_layer_loss(
+            cls_preds=cls_preds, box_cls_labels=box_cls_labels, num_class=self.num_class
+
+        )
```
Comment (lucagrementieri) [inline pcdet/models/dense_heads/anchor_head_template.py:107]: This function is calling itself if I'm not mistaken. How could that work?

Maybe you wanted to call `get_cls_layer_loss_one_set`?

File: pcdet/models/dense_heads/target_assigner/axis_aligned_target_assigner.py (hunk 6 above)
Comment (lucagrementieri) [reply pcdet/models/dense_heads/target_assigner/axis_aligned_target_assigner.py:58]: I don't know if a possible workaround could be to use
```
        gt_aux_classes = gt_boxes_with_classes[:, :, -1]
```

In that way the variable is always valid and if there's one less class, the content is just a repetition of the first class. In general this trick could allow for variable input sizes in this specific case of 1 or 2 labels.

File: setup.py (hunk 7 above)
Comment (lucagrementieri) [reply setup.py:31]: I think we can do a major version bump: 0.7.0.

**Review by lucagrementieri — APPROVED** [review]
The code is much better now, the edits are contained and very understandable.

I still have a suggestion to maybe increase the compatibility with the previous code, please look into that before mergin.

File: pcdet/models/dense_heads/anchor_head_template.py (hunk 3 above)
Comment (lucagrementieri) [reply pcdet/models/dense_heads/anchor_head_template.py]: Much better now!

File: pcdet/models/dense_heads/anchor_head_template.py (hunk 8 above)
Comment (julianschoep) [reply pcdet/models/dense_heads/anchor_head_template.py:107]: Oh damn good catch

File: pcdet/models/dense_heads/point_head_template.py
Comment (julianschoep) [reply pcdet/models/dense_heads/point_head_template.py:1]: It is defined in the data loader. This repo has a bunch of documentation problems (no type hinting for instance), I'd rather not update all the doc-strings everywhere to give explanation of this number or pull improving this in scope for this PR. Likely in the future we'll have to switch to a new repo anyway.
You're right though ofcourse, that it would be better to have some documentation on the structure

File: pcdet/models/dense_heads/target_assigner/axis_aligned_target_assigner.py (hunk 6 above)
Comment (julianschoep) [reply pcdet/models/dense_heads/target_assigner/axis_aligned_target_assigner.py:58]: Ah yes nice idea, done

**Comment by lucagrementieri** [comment]
@julianschoep I think that this new version of OpenPCDet should be integrate into IOS in isolation, i.e. without any model update, to ensure that there are no regressions and that the current model can still be used with this new version.
